File: apps/backend/app/api/routers/guests.py
from fastapi import APIRouter, Depends, Query

from app.core.auth import require_user_id
from app.core.tenancy import assert_org_member, assert_org_role
from app.schemas.domain import CreateGuestInput, UpdateGuestInput
from app.services.audit import write_audit_log
from app.services.table_service import create_row, delete_row, get_row, list_rows, update_row

router = APIRouter(tags=["Guests"])


@router.get("/guests")
def list_guests(
    org_id: str = Query(...),
    limit: int = Query(100, ge=1, le=500),
    user_id: str = Depends(require_user_id),
) -> dict:
    assert_org_member(user_id, org_id)
    return {"data": list_rows("guests", {"organization_id": org_id}, limit=limit)}


@router.post("/guests", status_code=201)
def create_guest(payload: CreateGuestInput, user_id: str = Depends(require_user_id)) -> dict:
    assert_org_role(user_id, payload.organization_id, {"owner_admin", "operator"})
    created = create_row("guests", payload.model_dump(exclude_none=True))
    write_audit_log(
        organization_id=payload.organization_id,
        actor_user_id=user_id,
        action="create",
        entity_name="guests",
        entity_id=created.get("id"),
        before_state=None,
        after_state=created,
    )
    return created


@router.get("/guests/{guest_id}")
def get_guest(guest_id: str, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("guests", guest_id)
    assert_org_member(user_id, record["organization_id"])
    return record


@router.patch("/guests/{guest_id}")
def update_guest(guest_id: str, payload: UpdateGuestInput, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("guests", guest_id)
    assert_org_role(user_id, record["organization_id"], {"owner_admin", "operator"})
    updated = update_row("guests", guest_id, payload.model_dump(exclude_none=True))
    write_audit_log(
        organization_id=record.get("organization_id"),
        actor_user_id=user_id,
        action="update",
        entity_name="guests",
        entity_id=guest_id,
        before_state=record,
        after_state=updated,
    )
    return updated


@router.delete("/guests/{guest_id}")
def delete_guest(guest_id: str, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("guests", guest_id)
    assert_org_role(user_id, record["organization_id"], {"owner_admin", "operator"})
    deleted = delete_row("guests", guest_id)
    write_audit_log(
        organization_id=record.get("organization_id"),
        actor_user_id=user_id,
        action="delete",
        entity_name="guests",
        entity_id=guest_id,
        before_state=deleted,
        after_state=None,
    )
    return deleted

File: apps/backend-rs/src/routes/guests.rs
use axum::{
    extract::{Path, Query, State},
    http::HeaderMap,
    response::IntoResponse,
    Json,
};
use serde_json::{json, Map, Value};

use crate::{
    auth::require_user_id,
    error::{AppError, AppResult},
    repository::table_service::{create_row, delete_row, get_row, list_rows, update_row},
    schemas::{
        clamp_limit, remove_nulls, serialize_to_map, validate_input, CreateGuestInput, GuestPath,
        GuestsQuery, UpdateGuestInput,
    },
    services::audit::write_audit_log,
    state::AppState,
    tenancy::{assert_org_member, assert_org_role},
};

pub fn router() -> axum::Router<AppState> {
    axum::Router::new()
        .route(
            "/guests",
            axum::routing::get(list_guests).post(create_guest),
        )
        .route(
            "/guests/{guest_id}",
            axum::routing::get(get_guest)
                .patch(update_guest)
                .delete(delete_guest),
        )
}

async fn list_guests(
    State(state): State<AppState>,
    Query(query): Query<GuestsQuery>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    assert_org_member(&state, &user_id, &query.org_id).await?;
    let pool = db_pool(&state)?;

    let mut filters = Map::new();
    filters.insert(
        "organization_id".to_string(),
        Value::String(query.org_id.clone()),
    );
    let rows = list_rows(
        pool,
        "guests",
        Some(&filters),
        clamp_limit(query.limit),
        0,
        "created_at",
        false,
    )
    .await?;
    Ok(Json(json!({ "data": rows })))
}

async fn create_guest(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(payload): Json<CreateGuestInput>,
) -> AppResult<impl IntoResponse> {
    validate_input(&payload)?;
    let user_id = require_user_id(&state, &headers).await?;
    assert_org_role(
        &state,
        &user_id,
        &payload.organization_id,
        &["owner_admin", "operator"],
    )
    .await?;
    let pool = db_pool(&state)?;
    let record = remove_nulls(serialize_to_map(&payload));
    let created = create_row(pool, "guests", &record).await?;
    let entity_id = value_str(&created, "id");
    write_audit_log(
        state.db_pool.as_ref(),
        Some(&payload.organization_id),
        Some(&user_id),
        "create",
        "guests",
        Some(&entity_id),
        None,
        Some(created.clone()),
    )
    .await;
    Ok((axum::http::StatusCode::CREATED, Json(created)))
}

async fn get_guest(
    State(state): State<AppState>,
    Path(path): Path<GuestPath>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;
    let row = get_row(pool, "guests", &path.guest_id, "id").await?;
    let org_id = value_str(&row, "organization_id");
    assert_org_member(&state, &user_id, &org_id).await?;
    Ok(Json(row))
}

async fn update_guest(
    State(state): State<AppState>,
    Path(path): Path<GuestPath>,
    headers: HeaderMap,
    Json(payload): Json<UpdateGuestInput>,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;
    let record = get_row(pool, "guests", &path.guest_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin", "operator"]).await?;
    let patch = remove_nulls(serialize_to_map(&payload));
    let updated = update_row(pool, "guests", &path.guest_id, &patch, "id").await?;
    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "update",
        "guests",
        Some(&path.guest_id),
        Some(record),
        Some(updated.clone()),
    )
    .await;
    Ok(Json(updated))
}

async fn delete_guest(
    State(state): State<AppState>,
    Path(path): Path<GuestPath>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;
    let record = get_row(pool, "guests", &path.guest_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin", "operator"]).await?;
    let deleted = delete_row(pool, "guests", &path.guest_id, "id").await?;
    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "delete",
        "guests",
        Some(&path.guest_id),
        Some(deleted.clone()),
        None,
    )
    .await;
    Ok(Json(deleted))
}

fn db_pool(state: &AppState) -> AppResult<&sqlx::PgPool> {
    state.db_pool.as_ref().ok_or_else(|| {
        AppError::Dependency(
            "Supabase database is not configured. Set SUPABASE_DB_URL or DATABASE_URL.".to_string(),
        )
    })
}

fn value_str(row: &Value, key: &str) -> String {
    row.as_object()
        .and_then(|obj| obj.get(key))
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(ToOwned::to_owned)
        .unwrap_or_default()
}
